Migrate modals from bootstrap-vue to vue-final-modal

jovyntls opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] Others, please explain:

Part of #1702 

**Overview of changes:**
Updates modals to use the [vue-final-modal](https://vue-final-modal.org/) package for modals instead of bootstrap-vue. 
- the modal transition CSS classes are moved to from markbind.css to Modal.vue with more specific selectors
- `bootstrapVueProcessor.js` is fully removed
	- the `_renameSlot` method is moved to `renameSlot` in `vueSlotSyntaxProcessor.js`
- Vue was upgraded from v2.6.11 to v2.6.14 to meet the dependencies of the latest version of vue-final-modal (v2.4.1)

DIscussion on using vue-final-modal and removing bootstrap-vue can be found on #1702.

**Anything you'd like to highlight / discuss:**
Deprecation of `modal-footer` and `modal-header` slots: 
The docs state that these slot are deprecated their implementations had not been removed. 
It seems like deprecating these slots may be an unrelated change, hence these slots are re-implemented in Modals and can be removed in a future PR for fully deprecating modal-header and modal-footer.

**Testing instructions:**
The existing documented attributes for modals should work [as described](https://markbind.org/userGuide/components/popups.html#modals).

**Proposed commit message: (wrap lines at 72 characters)**
```
Use vue-final-modal for modal components

Modals are implemented using bootstrap-vue directives.

bootstrap-vue has dependencies on specific versions of Bootstrap and
Vue, making it hard for MarkBind to migrate to Bootstrap 5, or Vue 3.

vue-final-modal has minimal dependencies, and supports both Vue 2 and 3.

Let's use vue-final-modal for modals instead. Doing so allows for future
migrations to Bootstrap 5 or Vue 3, without having to wait for
bootstrap-vue to migrate first. 

Vue is also upgraded from v2.6.11 to v2.6.14, to meet the dependencies
of vue-final-modal v2.4.1.
```

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

jonahtanjz commented on packages/vue-components/package.json:
```diff
@@ -41,6 +41,7 @@
     "jest": "^27.5.1",
     "jest-serializer-vue": "^2.0.2",
     "portal-vue": "^2.1.7",
+    "vue-final-modal": "2.3.2",
```
Just checking on this, I noticed that their latest version for Vue 2 is [`2.4.1`](https://www.npmjs.com/package/vue-final-modal). Is there a reason for using this specific version? 

Asking this because I saw in their [`2.4.0` release](https://github.com/vue-final/vue-final-modal/releases/tag/v2.4.0), they added the feature to import component directly. For example,
```js
import { vfmPlugin, $vfm, VueFinalModal } from 'vue-final-modal'
``` 
This should theoretically reduce the size of the bundle as only the required components are imported.

jovyntls replied:
Vue Final Modal (VFM) v2.4.1 requires Vue v2.6.14 as a dependency. However, upgrading Vue from v2.6.11 (current) to v2.6.14 causes hydration errors in Navbars. The hydration errors should be unrelated to this issue - they are present even when upgrading on the master branch.
The latest version of VFM compatible with Vue v2.6.11 is VFM v2.3.2.

jonahtanjz replied:
> The latest version of VFM compatible with Vue v2.6.11 is VFM v2.3.2.

Ahh ok got it :)

> However, upgrading Vue from v2.6.11 (current) to v2.6.14 causes hydration errors in Navbars. The hydration errors should be unrelated to this issue - they are present even when upgrading on the master branch.

Have you tried updating the dev version specified in [Site/index.js](https://github.com/MarkBind/markbind/blob/master/packages/core/src/Site/index.js#L319) to v2.6.14 as well to see if the hydration issue is still present? Might be cause by a mismatch of Vue's version. Will also need to update the [vue.min.js](https://github.com/MarkBind/markbind/blob/master/packages/core-web/asset/js/vue.min.js), which is used in production.

jonahtanjz commented on packages/vue-components/src/Modal.vue:
```diff
@@ -0,0 +1,155 @@
+  },
+};
+</script>
+<style scoped>
```
In view of #1768, we may want to consider using global styles instead since scoping may not be necessary here as these classes only appear in this file :)

jonahtanjz reviewed (commented):
Nice work @jovyntls 👍 Really clean implementation and good job on writing the tests. Will be nice to also include [Vue's snapshot test](https://github.com/MarkBind/markbind/tree/master/packages/vue-components/src/__tests__) for the newly added Modal component.

Just 1 suggestion, other than that the rest looks good :)

jovyntls commented:
Thanks for pointing it out @jonahtanjz , didn't realise the snapshot tests were a thing! I've changed the styles to global styles and added snapshot tests :)

jonahtanjz reviewed (approved):
Thanks @jovyntls! The tests look comprehensive :)

LGTM 👍

Will close and reopen this PR to restart the CI, looks like some issues with github recently 😅

ang-zeyu commented on packages/vue-components/src/Modal.vue:
```diff
@@ -0,0 +1,155 @@
+  <vue-final-modal
+    v-if="isMounted"
+    v-model="show"
+    ssr
```
Just to check, was `ssr=false` tried?

Presumably it prevents the modal from getting rendered during the ssr cycles, but I have been getting errors with it (might be a bug over there). The `isMounted` way seems like a good alternative.

Should we set `ssr=false` for clarity? (shouldn't be rendered during ssr) (maybe with a comment as well)

https://github.com/vue-final/vue-final-modal/blob/master/lib/VueFinalModal.vue#L3

jovyntls replied:
Yup! When `isMounted` is used, both `ssr` and `ssr=false` work. (Without `isMounted`, both don't work.) I've added a comment to clarify.

Just to check my understanding, why should modals not be rendered during SSR cycles? I used browser devtools to check for SSR hydration issues, not sure if there's any other way to detect SSR issues.

ang-zeyu replied:
> EDIT: ssr=false breaks the snapshot tests, so I reverted the change

Ok, ugh, the issue seems to be that `$vfm` (being a singleton) isn't really testable. Wonder if we should transfer this once we have e2e tests setup in the other PR. @jonahtanjz  🤔

Ok with snapshot tests now though. Let's update the comment as appropriate too.

> Just to check my understanding, why should modals not be rendered during SSR cycles? I used browser devtools to check for SSR hydration issues, not sure if there's any other way to detect SSR issues.

That depends, in MarkBind's case we've decided to discard modals from indexing, per the updates you did in `algolia.js`.

For this implementation https://github.com/vue-final/vue-final-modal/blob/master/lib/VueFinalModal.vue#L3, the `ssr` option seems to render a `display: none` modal. This might be useful if you want to include modal content inside search engine results.

Another case is if you want to show the modal on-page-load (rarely, since popups are triggered by user action usually), you would need to render it during ssr as well. (by setting the initial value of `show=true`)

ang-zeyu replied:
> per the updates you did in algolia.js.

EDIT on this: sorry, just realised with this discussion these selectors have to be retargeted / no longer necessary for SSR, something we missed out in #1534. (same rationale of excluding from indexing still applies though)

I've added it on to #1815.

jovyntls replied:
I see, thank you so much for the explanation!
Have updated the comment too :)

jonahtanjz replied:
> Wonder if we should transfer this once we have e2e tests setup in the other PR. @jonahtanjz 🤔

I think we can keep the current snapshot tests for now since the e2e tests are focused on the live preview part (this may not be relevant there?). We can consider moving this once we have more e2e tests for the other frontend components as well.

ang-zeyu reviewed (commented):
Nice work, just 2 nits:

---

The number of package-lock.json changes introduced in merging conflicts https://github.com/MarkBind/markbind/pull/1849/commits/c3e678f4690cc75204d421ed4f7ab7bc1df94f0f seems odd (compared with the earlier 2 commits changing package-lock.json)

jovyntls commented:
I tried checking out package-lock.json to the latest commit on master and then `npm run setup`, but it still gives 48k insertions and 25k deletions. O.o Not sure why there is such a large change since the only changes to the package.json files are adding `vue-final-modal` and upgrading the patch version of `vue` (along with `vue-server-renderer`, `vue-template-compiler`). 

Would this be an issue for the version history of package-lock.json?

ang-zeyu commented:
> I tried checking out package-lock.json to the latest commit on master and then `npm run setup`, but it still gives 48k insertions and 25k deletions. O.o Not sure why there is such a large change since the only changes to the package.json files are adding `vue-final-modal` and upgrading the patch version of `vue` (along with `vue-server-renderer`, `vue-template-compiler`).
> 
> Would this be an issue for the version history of package-lock.json?

Hmm, just tried a fresh (wiped all `node_modules` -- could try this) install, I'm not getting any updates on the main branch.

jovyntls commented:
> Hmm, just tried a fresh (wiped all `node_modules` -- could try this) install, I'm not getting any updates on the main branch.

I tried:
- restore `package-lock.json` to the version on the master branch
- delete all `node_modules`, including those in `packages/*` 
- `npm run setup`

Still getting the excessive changes, could be that I've messed up my config somehow, though I recently uninstalled and reinstalled the newest version of npm.

ang-zeyu commented:
👀 Just did a fresh clone + install on another os (ubuntu) but still not getting any changes.

Just in case, try `npm -V`. Might be an issue with multiple node/npm installations and your `PATH` configuration.

If it still persists think we could get @jonahtanjz to verify as well

jonahtanjz commented:
> If it still persists think we could get @jonahtanjz to verify as well

I'm able to get the correct changes in `package-lock.json` by restoring from the master branch and running `npm run setup`. I've put up the one I got on github [here](https://github.com/jonahtanjz/markbind/blob/vue-final-modal/package-lock.json). Maybe you can try replacing it with this and run `npm run setup` again to see if that helps?

jovyntls commented:
Ah ok, a fresh clone and re-installation worked for me! Not sure what the problem with the previous clone was.

Thank you @ang-zeyu and @jonahtanjz for the help :)
